### PR TITLE
Feature/kas 3756 beslissingen editor

### PR DESCRIPTION
**Pull request by kobemertens**
Needs custom backend for testing:
https://github.com/kanselarij-vlaanderen/app-kaleidos/tree/feature/KAS-3958-genereren-beslissingen

Also needs these 2 new services:
https://github.com/kanselarij-vlaanderen/decision-extraction-service
https://github.com/kanselarij-vlaanderen/decision-report-generation-service

JIRA tickets:
- https://kanselarij.atlassian.net/browse/KAS-3958
- https://kanselarij.atlassian.net/browse/KAS-3755
- https://kanselarij.atlassian.net/browse/KAS-3756

### CODE REVIEW

File: app/components/agenda/agendaitem/agendaitem-decision.js [hunk 1]
```diff
@@ -4,125 +4,436 @@ import { action } from '@ember/object';
...
 
   constructor() {
     super(...arguments);
     this.loadReport.perform();
     this.loadCodelists.perform();
+    this.loadNota.perform();
+  }
+
+  loadNota = task(async () => {
+    const nota = await this.agendaitemNota.notaOrVisieNota(
```
Comment (sergiofenoll) [inline app/components/agenda/agendaitem/agendaitem-decision.js]: Double check if visie nota is acceptable. Otherwise maybe extend the service you're calling here or make the "default to visie nota behaviour" configurable through a param.

File: app/components/agenda/agendaitem/agendaitem-decision.hbs [hunk 2]
```diff
@@ -1,69 +1,183 @@
...
-  </Auk::Panel::Body>
+  {{else}}
+    <Auk::Panel::Body
+      class="au-o-flow"
+    >
+      <Auk::Loader />
+    </Auk::Panel::Body>
+  {{/if}}
+
+  {{#if this.isEditing}}
```
Comment (sergiofenoll) [inline app/components/agenda/agendaitem/agendaitem-decision.hbs]: It would maybe be more logical to group all the "editing" blocks together instead of scatter them over multiple ifs. Not super important, we can revisit this later.

File: app/components/agenda/agendaitem/agendaitem-decision.hbs [hunk 3]
```diff
@@ -1,69 +1,183 @@
...
+    <Auk::Panel::Footer class="au-u-background-gray-100">
+      <Auk::Toolbar>
```
Comment (sergiofenoll) [inline app/components/agenda/agendaitem/agendaitem-decision.hbs]: We try to use `AuToolbar` now instead of `Auk::Toolbar`. Also not super important, but we can revisit this later.

File: app/components/agenda/agendaitem/agendaitem-decision.js [hunk 4]
```diff
@@ -4,125 +4,436 @@ import { action } from '@ember/object';
...
+      }
+    }
+    this.toggleEditPill();
+  });
+
+  /**
+   * Deprecated but needed for backwards compat
+   */
+  @action
+  async attachNewReportVersionDepr(piece) {
```
Comment (sergiofenoll) [inline app/components/agenda/agendaitem/agendaitem-decision.js]: Maybe rename this to `attachNewReportVersionAsPiece` and reword the comment so it explains that it's just the old way of working, i.e. a way to upload a new piece version for the report.

File: app/components/agenda/agendaitem/agendaitem-decision.js [hunk 5]
```diff
@@ -4,125 +4,436 @@ import { action } from '@ember/object';
...
+  exportPdf = task(async (report) => {
+    const resp = await fetch(`/generate-decision-report/${report.id}`);
+    if (!resp.ok) {
+      this.toaster.error(this.intl.t('error-while-exporting-pdf'));
+      return;
+    }
+    return await resp.json();
+  });
+
+  async replaceReportFile(report, fileId) {
```
Comment (sergiofenoll) [inline app/components/agenda/agendaitem/agendaitem-decision.js:189]: I would inline this because `deleteFile` immediately destroys the record, but here the report does not get immediately saved, so you need to be extra careful after calling this function to certainly call `report.save()`, otherwise you lose the link to the old file without replacing it with the new file.

Personal preference though, because at least in this class you do it correctly after calling the function.

File: app/components/agenda/agendaitem/agendaitem-decision.js [hunk 6]
```diff
@@ -4,125 +4,436 @@ import { action } from '@ember/object';
...
+    report.modified = new Date();
   }
 
   @action
-  toggleEdit() {
-    this.isEditing = !this.isEditing;
+  handleRdfaEditorInitBetreft(editorInterface) {
+    if (this.betreftPiecePart) {
+      editorInterface.setHtmlContent(this.betreftPiecePart.value);
+    } else if (this.editorInstanceBetreft) {
```
Comment (sergiofenoll) [inline app/components/agenda/agendaitem/agendaitem-decision.js:244]: This is from some possible buggy behaviour with re-rendering, maybe document it by linking to a rocket.chat conversation, if that's publicly available.

File: app/components/agenda/agendaitem/agendaitem-decision.js [hunk 7]
```diff
@@ -4,125 +4,436 @@ import { action } from '@ember/object';
...
   }
 
-  @task
-  *loadReport() {
-    this.report = yield this.args.decisionActivity.report;
-    this.previousReport = yield this.report?.previousPiece;
+  @action
+  updateBetreftContent() {
+    this.setBetreftEditorContent(
+      `<p>${this.args.agendaContext.agendaitem.shortTitle}</p>`
```
Comment (sergiofenoll) [inline app/components/agenda/agendaitem/agendaitem-decision.js]: Short title is only a part of the "betreft" piece-part. This also needs the long title of the agendaitem, as well as a list of the piece names with wrapping `()`.

This will have to be iterated on further down the line, but it should be okay to add those two for now.

File: app/components/agenda/agendaitem/decision-viewer.hbs
Comment (sergiofenoll) [inline app/components/agenda/agendaitem/decision-viewer.hbs:1]: Since we no longer use this, I would remove the component, but we still have it in the git history if we ever need it later.

File: app/helpers/date-weekday.js
Comment (sergiofenoll) [inline app/helpers/date-weekday.js:1]: This is technically unused if you end up removing `Agenda::Agendaitem::DecisionViewer`, but I'm okay with leaving it in since it seems useful in other contexts.

File: app/models/agendaitem.js [hunk 8]
```diff
@@ -27,7 +27,12 @@ export default class Agendaitem extends ModelWithModifier {
 
   @hasMany('mandatee', { inverse: 'agendaitems', async: true }) mandatees;
   @hasMany('piece', { inverse: 'agendaitems', async: true }) pieces;
```
Comment (sergiofenoll) [inline app/models/agendaitem.js]: Should this one also be `polymorphic`?

File: app/models/piece-part.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import Model, { attr, belongsTo } from '@ember-data/model';
+
+export default class PiecePart extends Model {
+  @attr('string') title;
+  @attr('string') value;
```
Comment (sergiofenoll) [inline app/models/piece-part.js]: Ok as-is, but if the type of the prop is `string` it's enough to just write `@attr prop`.

File: app/components/agenda/agendaitem/agendaitem-decision.hbs [hunk 10]
```diff
@@ -1,69 +1,183 @@
...
-      <Auk::Loader />
+      {{#if (and this.report.file (not this.report.pieceParts))}}
+        <Auk::Panel::Body class="au-o-flow" @noPadding={{true}}>
+          <Documents::DocumentCard
+            @piece={{this.report}}
+            @isEditable={{true}}
+            @agendaContext={{@agendaContext}}
+            @hideNewerVersions={{true}}
+            @onOpenUploadModal={{@onOpenUploadModal}}
+            @onAddPiece={{this.attachNewReportVersionDepr}}
```
Comment (sergiofenoll) [inline app/components/agenda/agendaitem/agendaitem-decision.hbs]: `@onAddPiece` will pass a `piece` record to `this.attachNewReportVersionDepr`. We actually need a `report` record, otherwise ED cries. You'll need to find a way to cast it/convert it/create a new report with the same attributes as the piece and delete the piece.

Not sure if there's a cleaner way, maybe parametrize the document-card so you can tell which kind of piece to create.

File: app/styles/app.scss
Comment (sergiofenoll) [inline app/styles/app.scss:1]: I think this can go, because the we no longer show the decision viewer.

File: app/styles/layout/_decision-viewer.scss
Comment (sergiofenoll) [inline app/styles/layout/_decision-viewer.scss:1]: Can be removed, decision viewer is no longer shown.

File: public/assets/logo_vlaamse_regering.svg
Comment (sergiofenoll) [inline public/assets/logo_vlaamse_regering.svg:1]: I don't know a lot about SVG, but since the changed version is no longer necessary because we're no longer showing the decision in the app, maybe undo this change in case it breaks anything elsewhere.

Though it may very well be a completely safe change.

File: translations/nl-be.json [hunk 11]
```diff
@@ -1182,8 +1184,25 @@
...
+  "concerns": "Betreft",
+  "update-thing": "{thing} updaten",
+  "current-version": "Huidige versie",
+  "monday": "maandag",
+  "tuesday": "dinsdag",
+  "wednesday": "woensdag",
+  "thursday": "donderdag",
+  "friday": "vrijdag",
+  "saturday": "zaterdag",
+  "sunday": "zondag",
```
Comment (sergiofenoll) [inline translations/nl-be.json]: Not sure, but I think `date-fns` has a way to display the weekday name according to the app's locale (`nl-be`). Maybe look into that to see if its the case? https://date-fns.org/

**Review by sergiofenoll — CHANGES_REQUESTED** [review]

File: app/components/agenda/agendaitem/agendaitem-decision.js (hunk 5 above)
Comment (kobemertens) [reply app/components/agenda/agendaitem/agendaitem-decision.js:189]: I added a `report.save()` to the end of the function, I prefer having a descriptive name for the function over performance in this case.

File: app/components/agenda/agendaitem/agendaitem-decision.js (hunk 1 above)
Comment (kobemertens) [reply app/components/agenda/agendaitem/agendaitem-decision.js]: Should only be nota, fixed in: [545ae4d](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/545ae4db156308d172c9f4b7bbfa7cab34798fdb)

File: app/components/agenda/agendaitem/agendaitem-decision.hbs (hunk 2 above)
Comment (kobemertens) [reply app/components/agenda/agendaitem/agendaitem-decision.hbs]: fixed in [e2e5e9d](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/e2e5e9d0597e7746a3f44474a2175525ea3c49a1)

File: app/components/agenda/agendaitem/agendaitem-decision.js (hunk 4 above)
Comment (kobemertens) [reply app/components/agenda/agendaitem/agendaitem-decision.js]: fixed in [e331baa](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/e331baabb8132ea4c05270437106f8f3fb6fe87f)

File: app/components/agenda/agendaitem/agendaitem-decision.js (hunk 6 above)
Comment (kobemertens) [reply app/components/agenda/agendaitem/agendaitem-decision.js:244]: linked to conversation in the code. 
https://chat.semte.ch/channel/say-editor?msg=q9gF5BfAHFWiyGv84

File: app/components/agenda/agendaitem/agendaitem-decision.js (hunk 7 above)
Comment (kobemertens) [reply app/components/agenda/agendaitem/agendaitem-decision.js]: Added in [b1e5142](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/b1e5142c38932725712fac79eef2399866ceb906)

File: app/components/agenda/agendaitem/decision-viewer.hbs
Comment (kobemertens) [reply app/components/agenda/agendaitem/decision-viewer.hbs:1]: removed in [62ec60d](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/62ec60df7f42e19a2439a7c9c111605070b653e5)

File: app/helpers/date-weekday.js
Comment (SchrodingersCat00) [reply app/helpers/date-weekday.js:1]: removed in [62ec60d](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/62ec60df7f42e19a2439a7c9c111605070b653e5)

File: app/models/piece-part.js (hunk 9 above)
Comment (kobemertens) [reply app/models/piece-part.js]: removed in [f692371](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/f692371ef9fb3f6d338c2e02f5c89ba7fa68bc5b)

File: app/components/agenda/agendaitem/agendaitem-decision.hbs (hunk 10 above)
Comment (kobemertens) [reply app/components/agenda/agendaitem/agendaitem-decision.hbs]: added parameter to document card in [23de5e7](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/23de5e762aba346097f18d6d961731132bd31972)

File: app/models/agendaitem.js (hunk 8 above)
Comment (kobemertens) [reply app/models/agendaitem.js]: added in [e63d592](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/e63d592ec0c1245f1ef63097fc7a6e1ce782bfdc)

File: public/assets/logo_vlaamse_regering.svg
Comment (kobemertens) [reply public/assets/logo_vlaamse_regering.svg:1]: reverted in [62ec60d](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/62ec60df7f42e19a2439a7c9c111605070b653e5)

File: translations/nl-be.json (hunk 11 above)
Comment (kobemertens) [reply translations/nl-be.json]: implemented in [5dbecf7](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/5dbecf7cb44370e6f771d6cff79e0ab931be8377)

File: app/styles/layout/_decision-viewer.scss
Comment (kobemertens) [reply app/styles/layout/_decision-viewer.scss:1]: removed in [62ec60d](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/62ec60df7f42e19a2439a7c9c111605070b653e5)

File: app/styles/app.scss
Comment (kobemertens) [reply app/styles/app.scss:1]: removed in [62ec60d](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/62ec60df7f42e19a2439a7c9c111605070b653e5)

**Review by sergiofenoll — APPROVED** [review]

**Comment by sergiofenoll** [comment]
Approved but let's wait on merging this until we figure out how we want to deal with supporting the old workflow

**Comment by kobemertens** [comment]
[32659bd](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/pull/1713/commits/32659bde06caaedc5ea5234be9f9f800990bf1ba) adds a feature flag to enable the new UI. This should be ready to be merged

File: app/components/agenda/agendaitem/agendaitem-decision.js [hunk 12]
```diff
@@ -2,69 +2,199 @@ import Component from '@glimmer/component';
...
+    }
+    return await resp.json();
+  });
+
+  async replaceReportFile(report, fileId) {
+    await deleteFile(report.file);
+    const file = await this.store.findRecord('file', fileId);
+    report.file = file;
+    report.modified = new Date();
+    report.save();
```
Comment (kobemertens) [inline app/components/agenda/agendaitem/agendaitem-decision.js]: this needs to be `await`ed

**Review by ValenberghsSven — APPROVED** [review]
Reviewed/resolved comments in pair meeting.
Changed tests to work with the new way of creating decisions.
Disabled decision.spec because it was too much work to get everything working right now + We should keep testing the "legacy" decisions (even though ideally there should be no changes on existing documents on PROD, so mainly focus on correct viewing for all profiles)
